Clean up uses of deprecated API's in node core

watson opened the pull request:
## Summary

Ensure no deprecated Node.js core API's are used in Kibana. This is achieved by throwing an error in either development mode or in CI if one of the deprecated API's is called, and as such, new PR's should no longer be able to be merged if they use deprecated API's.

Some of these API's (like the `Buffer` constructor`) is a security risk.

### For reviewers

A few of the existing tests have been skipped. This is because the tests were actually not running correctly, but contained errors that were silently ignored because an error was thrown inside of a promise which was then unhandled.

In most of these cases, the issue is that some of the API was mocked, but the mocked object didn't contain the expected properties, and the mock needs to be updated. I'm not the person best equipped to make these updates, so I think we should just merge this PR as is, and create issues to get the skipped tests re-added.

If you'd like to run one of the skipped jest or mocha tests locally to see more details about the error, do the following:

1. Change `skip` to `only` inside the file
1. Add the following snippet in the top of the file:
    ```js
    process.on('unhandledRejection', reason => {
      process._rawDebug('Unhandled Rejection:', reason);
    });
    ```
1. Run the file like so:
    ```
    node scripts/jest <path-to-file>
    # or:
    node scripts/mocha <path-to-file>
    ```
    _If the file is within the `x-pack` directory, first `cd x-pack` then run the same command as above, but leave out the `x-pack/` prefix in the path to the file._

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] Run test suite with `--throw-deprecation` to find the uses of deprecated API's and fix those until all have been fixed.
- [x] Consider if we should run with `--throw-deprecation` in CI in the future, or only `--trace-deprecation`
- [x] Add `--throw-deprecation` to Kibana dev mode
- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
~~- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

watson committed:
tmp: step 1, figure out where deprecated API's are used

watson committed:
Use non-deprecated Buffer API

watson committed:
pretier

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11521/)
* Commit: 9ea06cf766232d7e4f1b06e836c459d8d413c55d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11527/)
* Commit: ff823a69d3ba257931794a5e935b125ad5facd61

watson committed:
fix yaxis error

watson commented on src/legacy/core_plugins/timelion/server/series_functions/__tests__/yaxis.js:
```diff
@@ -95,16 +95,16 @@ describe('yaxis.js', () => {
   it('throws an error if currency is not three letter code', () => {
     invoke(fn, [seriesList, 1, null, null, null, null, null, 'currency:abcde']).catch(e => {
-      expect(e).to.be.an(Error);
+      expect(e).to.be.an.instanceof(Error);
```
These tests were actually not running correctly and just silently failed without breaking the test itself. It triggered an exception that was caught by the outer `catch` function which then logged an "unhandled promise rejection" desperation warning, which in turn went unnoticed.

watson commented:
FYI @elastic/kibana-operations

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11560/)
* Commit: 1cf4d8068be2f362b11bd0bbae6132caf19f0ce2

watson commented:
I wanted to enable `--pending-deprecation` in development as well, but this currently will make it impossible to start Kibana as we're dependent on an old version of `convert-source-map` which is using the un-safe Buffer API. This dependency comes from `@babel/core` which is also an old version. Once we upgrade that this should go away and we can enable `--pending-deprecation` as well.

watson committed:
Hide unhandled promise rejection in test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12159/)
* Commit: 8100f2c3d6030791c648eafa3900375bb426fd02

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12179/)
* Commit: 07696df0a0bc318e07a8a6c27695a1e5209d5e34

watson committed:
use pixelmatch instead of image-diff

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12241/)
* Commit: 7db6e6526d97ea8cd1c549fb961670d4699c08d9

watson committed:
fix more Buffer

watson committed:
bump pixelmatch in x-pack

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12272/)
* Commit: 185018360177b6bfe34d14ca121d1413c77d4ab8

watson committed:
Ensure promises are not unresolved

watson committed:
fix index_management test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12376/)
* Commit: 350458d6d7424e88a53f506add69ad7b453671d2

watson committed:
fix monitoring test

watson committed:
add TODO's

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12457/)
* Commit: 618f88f6755d05da1c1c2b54692810d01a259868

watson committed:
fix create_index_pattern_wizard tests

watson committed:
skip broken saved_objects tests

watson committed:
skip broken query_bar tests

watson committed:
skip broken query_bar_input tests

watson committed:
skip broken dashboard_grid tests

watson committed:
skip broken dashboard_viewport tests

watson committed:
skip broken replace_panel_action tests

watson committed:
skip broken apply_filter_action test

watson committed:
skip broken x-pack index_privileges test

watson committed:
throw on deprecations in development mode

watson committed:
remove changes unrelated to the PR

watson committed:
skip broken x-pack security test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12599/)
* Commit: 7531eaf73519826615694a9e0e33c7e6c6ca3ade

watson committed:
fix error in reporting into button React component

watson commented on src/legacy/core_plugins/data/public/query/query_bar/components/query_bar_top_row.test.tsx:
```diff
@@ -121,7 +121,7 @@ describe('QueryBarTopRowTopRow', () => {
   });
 
-  it('Should render the given query', () => {
+  it.skip('Should render the given query', () => {
```
@lizozom I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson commented on src/legacy/core_plugins/data/public/query/query_bar/components/query_string_input.test.tsx:
```diff
@@ -90,7 +90,7 @@ function wrapQueryStringInputInContext(testProps: any, storage?: any) {
 }
 
-describe('QueryStringInput', () => {
+describe.skip('QueryStringInput', () => {
```
@lizozom I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12622/)
* Commit: 7002751ec18f283118a4b69027db4d8803dc80b3

watson commented on src/plugins/dashboard_embeddable_container/public/actions/replace_panel_action.test.tsx:
```diff
@@ -79,7 +79,7 @@ beforeEach(async () => {
 });
 
-test('Executes the replace panel action', async () => {
+test.skip('Executes the replace panel action', async () => {
```
@streamich I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson commented on src/plugins/dashboard_embeddable_container/public/embeddable/grid/dashboard_grid.test.tsx:
```diff
@@ -93,7 +93,7 @@ afterAll(() => {
 });
 
-test('renders DashboardGrid', () => {
+test.skip('renders DashboardGrid', () => {
```
@streamich I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson commented on src/plugins/dashboard_embeddable_container/public/embeddable/viewport/dashboard_viewport.test.tsx:
```diff
@@ -88,7 +88,7 @@ function getProps(
 }
 
-test('renders DashboardViewport', () => {
+test.skip('renders DashboardViewport', () => {
```
@streamich I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson commented on src/plugins/embeddable/public/lib/actions/apply_filter_action.test.ts:
```diff
@@ -93,7 +93,7 @@ const getEmbeddable = () => {
 describe('execute()', () => {
   describe('when no filters are given', () => {
-    test('throws an error', async () => {
+    test.skip('throws an error', async () => {
```
@streamich I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson replied:
Here's the stack trace of the unhandled promise rejection:

```
Error: Applying a filter requires a filter and embeddable as context
    at Object.execute (./src/plugins/embeddable/public/lib/actions/apply_filter_action.ts:58:15)
    at error (./src/plugins/embeddable/public/lib/actions/apply_filter_action.test.ts:104:16)
    at expectError (./src/plugins/embeddable/public/tests/helpers.ts:28:5)
    at Object.test.only (./src/plugins/embeddable/public/lib/actions/apply_filter_action.test.ts:103:21)
    at Object.asyncJestTest (./node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at resolve (./node_modules/jest-jasmine2/build/queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (./node_modules/jest-jasmine2/build/queueRunner.js:26:19)
    at promise.then (./node_modules/jest-jasmine2/build/queueRunner.js:73:41)
```

I don't see a reason to try and fix this before this PR is merged as the test fails already on the 2nd line, but the failure is just suppressed due to the unhandled rejection not being taken seriously.

stacey-gammon replied:
Yea, this is pretty funny, the `expectError` function is wrong - at least it always returns an Error object no matter what.  I have a fix for this test and the function here: https://github.com/elastic/kibana/pull/52078 . Since I changed the fn it may end up having to be a bigger PR... waiting to see what ci tells me.  cc @streamich 

My concern with skipping is that all these tests will be forgotten about, like the Embeddables which has been turned off for four months in order to get a PR merged (I can't even find the issue now...). I do see your point about the tests being useless now though.  We really just need a better way to solve the issue with skipped tests being forgotten about and not prioritized.

watson replied:
This test is "useless" in the sense it doesn't check anything because it's currently aborted before it even starts to assert anything - and so skipping it doesn't reduce test coverage.

No one is arguing that we should leave tests skipped. Skipped tests should normally be avoided like the plague. But the pros of merging this PR with this test skipped far outweigh the cons as we currently can't trust any of our tests with `async` functions. This, in turn, might introduce even bigger and more serious errors as a result. And a skipped test is at least more visible than a green test that silently fails.

As described in the PR description, I'll make issues and assign them to the appropriate teams to get the test un-skipped once this PR lands. But if you feel like fixing this bug already now, please push your fixes to this PR to avoid merge conflicts and so this PR isn't blocked.

stacey-gammon replied:
> This test is "useless" in the sense it doesn't check anything because it's currently aborted before it even starts to assert anything - and so skipping it doesn't reduce test coverage.

Yes, I understand, it's useless in its *current form*, didn't mean to imply otherwise. 

Perhaps I was unclear, my comment about skipped tests was just a comment about concern with them in general, not a decision on the PR (and actually, it seems like the team has been doing well fixing `skipped-test` issues - only a couple outstanding, so maybe my concern was wrong).  I am fine if you merge this PR with these kinds of tests skipped - ones where skipping doesn't actually reduce test coverage. I only looked at this one test though - @streamich mentioned some other skipped ones did still provide test coverage.  It was my understanding you and he synced after the PR description was written and the agreed upon outcome was that the app arch team would fix the tests first, before the PR was merged.  It sounds like that perhaps there wasn't actually a clear understanding of that.

As far as that other PR goes - I don't think it should go in this one because it might grow in scope if I hit PR failures (I suspect it will since I fixed the `expectError` fn).  By keeping it separate, you can merge before I finish that PR. Keeping them separate actually helps not block your PR, imo.  Also, it shouldn't cause merge conflicts - we haven't edited any of the same lines. If I merge that PR first, you'd just have to undo the skip. If you merge first, I'll undo the skip.  Thoughts?

stacey-gammon replied:
As it turns out, that fix was easier than I thought, required no other changes.  I'd still like to go ahead and merge that separately.  Then that test fix won't be held up by this PR.  Like I said though, it really shouldn't cause any merge conflicts since all you touched in here was adding the `skip`.  If it does cause an issue though, lmk, I'll help resolve.

watson replied:
Good to hear it was an easy fix. I'll remove the `skip` and rebase with `master` to get your fix in.

watson replied:
Rebased 👍

watson commented on src/plugins/kibana_react/public/saved_objects/saved_object_finder.test.tsx:
```diff
@@ -248,7 +248,7 @@ describe('SavedObjectsFinder', () => {
     });
 
-    it('should include additional fields in search if listed in meta data', async () => {
+    it.skip('should include additional fields in search if listed in meta data', async () => {
```
@flash1293 I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson replied:
Here's the stack trace of the unhandled promise rejection:

```
TypeError: Cannot read property 'savedObjects' of undefined
    at SavedObjectFinder._lodash.default.debounce (./src/plugins/kibana_react/public/saved_objects/saved_object_finder.tsx:46:32)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

flash1293 replied:
Should be fixed now, I pushed on this PR.

watson replied:
Looking good! Thanks for helping out and providing the fix

watson commented on utilities/visual_regression.js:
```diff
@@ -110,17 +108,23 @@ async function compareScreenshots() {
+    const { width, height } = sessionImage;
+    const diff = new PNG({ width, height });
+
+    const numDiffPixels = pixelmatch(
```
@cjcenizal I've replaced the `image-diff` module with `pixelmatch` as `image-diff` is deprecated and makes use of deprecated API's in Node core. However, I've not been able to verify if this new code works. As far as I can see this is used to create a "diff gallery" in HTML form. What's the best way for me to run this and see the output?

watson replied:
@elastic/kibana-qa Just realized that CJ is on leave currently. Do any of you know about this part of the code?

LeeDr replied:
FYI, I'm looking at test `dashboard_app_using_current_data_dashboard_snapshots_compare_TSVB_snapshot/`
in the Chrome UI Functional Tests

This looks correct (it's headless Chome so this is the simplest case);
```
[00:16:09]                 │ debg ======browser======== setWindowSize 1000 500
[00:16:09]                 │ debg ======browser======== actual initial screenshot size width=1200, height=800
[00:16:10]                 │ debg ======browser======== actual second screenshot size width= 600, height=400
[00:16:10]                 │ debg ======browser======== calculated values xBorder= 0, yBorder=0, xScaling=1, yScaling=1
[00:16:10]                 │ debg ======browser======== setting browser size to 1000 x 500
[00:16:10]                 │ debg ======browser======== final screenshot size width=1000, height=500
```

And the actual screenshot and comparison works correctly;
```
[00:16:43]                 │ info Taking screenshot "/dev/shm/workspace/kibana/test/functional/screenshots/session/tsvb_dashboard.png"
[00:16:43]                 │ debg comparePngs: /dev/shm/workspace/kibana/test/functional/screenshots/session/tsvb_dashboard.png vs /dev/shm/workspace/kibana/test/functional/screenshots/baseline/tsvb_dashboard.png
[00:16:44]                 │ debg calculating diff pixels...
[00:16:44]                 │ debg percent different: 0.012874
```

watson replied:
Phew! Glad that it worked - it's pretty rare that writing code without running it resolves in a functional program in the first go 😅Thanks for testing it.

watson commented on x-pack/legacy/plugins/reporting/public/components/report_info_button.tsx:
```diff
@@ -266,7 +266,6 @@ export class ReportInfoButton extends Component<Props, State> {
           info: null,
           error: kfetchError,
         });
-        throw kfetchError;
```
@tsullivan I've taken the liberty to remove this re-throw. As far as I can see, this just results in an unhandled promise rejection error - at least in the tests, so I assume in prod as well. And since we set the error state in the line above, I think we're ok by not re-throwing. But I'm not 100% sure, and since it seems that you touched this code last, I thought I'd just ping you if you don't mind.

tsullivan replied:
+1 it looks like we are already doing what we need to with this error, which is add it to the component state so it can appear in a message.

watson commented on x-pack/legacy/plugins/security/public/views/management/edit_role/components/privileges/es/index_privileges.test.tsx:
```diff
@@ -10,7 +10,7 @@ import { RoleValidator } from '../../../lib/validate_role';
 import { IndexPrivileges } from './index_privileges';
 
-test('it renders without crashing', () => {
+test.skip('it renders without crashing', () => {
```
@legrego I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson commented on x-pack/plugins/security/server/authentication/authenticator.test.ts:
```diff
@@ -515,7 +515,7 @@ describe('Authenticator', () => {
     });
 
-    it('only updates the session lifespan expiration if it does not match the current server config.', async () => {
+    it.skip('only updates the session lifespan expiration if it does not match the current server config.', async () => {
```
@jportner I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

jportner replied:
Yes, sounds like my fault, I will look into it, thanks!

dgieselaar reviewed (approved):
APM changes LGTM

watson committed:
remove debug code

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12657/)
* Commit: 2d26f96f215b721606b3205f4d08ec63cf3968a4

watson commented:
@elasticmachine merge upstream

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12804/)
* Commit: 9b0bc4733e5f9e50d5c8114728868aafbb5fcabe

jportner committed:
Fix authenticator.test.ts

Test was broken but it was not waiting for promise resolution.
I reworked the test into four separate sub-tests.

rudolf reviewed (approved):
:+1: for Platform changes

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12890/)
* Commit: f8e87afa9946e787d53421615f7f10fc3328fb92

jportner committed:
Fix index_privileges.test.tsx

One test had a false negative failure --
Fixed the test and also added checks to ensure promises resolve.

jportner commented on x-pack/legacy/plugins/security/public/views/management/edit_role/components/privileges/es/index_privileges.test.tsx:
```diff
@@ -10,59 +10,47 @@ import { RoleValidator } from '../../../lib/validate_role';
-      name: '',
-      elasticsearch: {
-        cluster: [],
-        indices: [],
```
@legrego -- passing in an empty `indices` array was causing a rejected promise. As a result, the component never mounted correctly in this test (see `index_privileges.test.tsx.snap`).

Instead of adding a check to `index_privileges.tsx` to change the behavior and allow that test to pass, I opted to change the props and pass in a non-empty `indices` array. I assume this is intended functionality, but please take a look and correct it if I made the wrong call.

legrego replied:
Thanks for taking a look Joe! I still think it's valuable to test that the component can render without any `indices` provided, as this is a common real-world use-case. Taking a look at the component itself, I think the offending code is here:

https://github.com/elastic/kibana/blob/0a527c658c50572d66a71a62cb47844edea7ab4a/x-pack/legacy/plugins/security/public/views/management/edit_role/components/privileges/es/index_privileges.tsx#L157-L164

Since we are calling `this.setState` in a callback, it is possible that we won't call this until after the component has unmounted. I believe that's what's happening in the test case. I think the "proper" fix is to track the mounted status of this component, and only call `this.setState` if the component is still mounted.

jportner replied:
OK, for future reference the problem was actually Enzyme not being able to handle a component that renders as an empty array or null. Wrapping the `index_privileges_form` component in a `Fragment` fixed the problem!

jportner reviewed (approved):
Security changes LGTM.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12966/)
* Commit: 56c31498f63f0db65a24b2fa9eb820208324d626

mistic reviewed (approved):
I really like this! The code owners changes for the operations team LGTM

streamich reviewed (changes requested):
Why do you want to skip the tests that report unhandled error exception? Unhandled error rejections could be in the code itself, not tests.

We could create an issue to fix the unhandled promise rejection warnings, but I'm not sure what we gain from disabling the tests. If we simply disable the tests somebody could introduce a regression.

Shall we rather for each plugin that causes unhandled promise rejection warnings create an issue to fix that?

watson commented:
@streamich I tried to address the issue of the skipped tests in the PR description under the "For reviewers" headline:

> [...] In most of these cases, the issue is that some of the API was mocked, but the mocked object didn't contain the expected properties, and the mock needs to be updated. I'm not the person best equipped to make these updates, so I think we should just merge this PR as is, and create issues to get the skipped tests re-added. [...]

For each of the tests that I skipped, I first tried to figure out why the test promise was not resolved and if I could fix the problem. In a few scenarios, the unresolved promise rejection was part of the test and I could fix those. But in most cases, the issue was either in the mocks (in which case the test didn't actually test anything anyway, and disabling doesn't make things worse), or the issue was in the code itself and the test actually found this issue, but failed to break because we didn't throw on unhandled rejections.

In any case, I think trying to fix all these tests etc in one PR is going to mean that it will take forever to merge and in the meantime new issues will pop up in `master`, which in turn will mean that this PR will need to be rebased and the process starts all over again.

I don't think that merging this PR will leave us any worse off than we already are, so I suggest that we merge it and then I'll make sure to create issues to get the skipped tests fixed accordingly.

But if you have identified a specific test that you would like to get fixed as part of this PR as opposed to in a followup PR, please let me know and we'll try to get the appropriate people involved 🙂

streamich commented:
I'm not suggesting to "fix" the tests in this PR. I was suggesting to re-enable the tests and create an issue with a list of tests/plugins that need to address unhandled promise rejection problem, so that those can be done by their respective owners.

If unhandled promise rejection is in the code, it does not mean the test is wrong. Also, if there is some unhandled error rejection in the code it does not mean that a test should break.

> I don't think that merging this PR will leave us any worse ...

It will leave us worse because we will have a bunch tests disabled, for example in dashboard, and somebody can introduce a regression.

watson commented:
> It will leave us worse because we will have a bunch tests disabled, for example in dashboard, and somebody can introduce a regression.

I don't think this is the case. The now skipped tests don't actually test anything in `master` today, so skipping them doesn't change the test coverage. Today they are all silently failing but reporting themselves as green because the promise rejection was unhandled.

streamich commented:
> The now skipped tests don't actually test anything in master today, so skipping them doesn't change the test coverage.

This is definitely not the case. Just tested a couple of the disabled tests and they do in fact work.

> Today they are all silently failing but reporting themselves as green because the promise rejection was unhandled.

You intention is good: to address unhandled promise rejection in *Node.js*. But the tests you are disabling don't test the *Node.js* code, they test browser code.

watson commented:
I'm not sure if you're only referring to the dashboard tests, but take the skipped test in `src/plugins/dashboard_embeddable_container/public/actions/replace_panel_action.test.tsx` for example. The unhandled promise rejection here comes from the following runtime exception:

```
TypeError: Cannot read property 'overlays' of undefined
    at openReplacePanelFlyout (./src/plugins/dashboard_embeddable_container/public/actions/open_replace_panel_flyout.tsx:43:30)
    at ReplacePanelAction.execute (./src/plugins/dashboard_embeddable_container/public/actions/replace_panel_action.tsx:84:59)
    at Object.<anonymous>.test.only (./src/plugins/dashboard_embeddable_container/public/actions/replace_panel_action.test.tsx:96:10)
    at Object.asyncJestTest (./node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at resolve (./node_modules/jest-jasmine2/build/queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (./node_modules/jest-jasmine2/build/queueRunner.js:26:19)
```

The line numbers are not quite right due to issues with multiple levels of transpilation, but if you search for `overlays` in the file mentioned in the top frame, there is only one place where this exception can come from:

https://github.com/elastic/kibana/blob/a80366b36481c8efb3046a7766e90fe241f1b5e3/src/plugins/dashboard_embeddable_container/public/actions/open_replace_panel_flyout.tsx#L47

So `core` in this case is `undefined`. I suspect this is because of a mock inside the test isn't set up correctly, but if you look at the 3rd frame in the stack trace, you'll see that it comes from within the test when the test tries to run `action.execute({ embeddable })`, so the test actually never gets to test anything as far as I can see.

majagrubic commented:
Please do not skip tests, as this is a dangerous practice. Kibana App Team is doing some major refactoring in several areas of the code at the moment and we rely on these tests to make sure things work properly. Skipping them silently will leave the team in a vulnerable position. I can help you look into this the failures introduced by this code, but please do not merge with skipped tests.

sqren commented:
@streamich @majagrubic AFAICT there is some misunderstanding about what this pull request does. Some tests were failing silently. Now @watson is enabling a stricter setting so they no longer fail silently. This leaves the codebase IN A MUCH BETTER STATE. I can't stress how important this change is. For this change to make it in, the failing tests have to be fixed or skipped. I don't think we should put the burden on @watson (eg. @jportner already stepped up and fixed one of the affected tests). 
I agree that skipping tests is mostly not a good solution but in this case it'll make the overall codebase so much better - and again, afaict most of the skipped tests were failing anyway (I didn't look closely into all of them so could be mistaken).

Commit:
Fixing failing dashboard_grid tests

Commit:
Fixing dashboard_viewport test

watson commented:
Yes as @sqren points out, _any_ of the tests that are using an `async` function can throw synchronously today and we will never know because it's caught by the outer promise which then emits the unhandled promise rejection warning - which in turn, this PR now handles in a more strict way.

So while you are all correct that _some_ of the tests runs to completion with all its assertions today, and that it's only _after_ a test finishes that the promise rejection is unhandled (and so skipping these tests leave us vulnerable to a regression where those tests would previously have covered us), the reality is, that without this PR, any of our `async` tests can throw midway through and hide an even more serous error. So I genereally think that the pros far outweight the cons if we were to merge this PR as is.

That being said, I'm super happy to have already gotten the help from both @jportner and @majagrubic who both pushed test fixes to this PR so that we can land it without those tests being skipped. **So if there's a "high risk" test you think is required to be fixed before this PR is merged, please make a comment in the code letting me know**, and if you have the expertise, please push a commit to this PR with a fix as well. I've already fixed the tests I could myself, so the ones that are left are unfortunately outside my area of expertise - so any help here is highly appreciated.

I'll make sure to create new issues related to any skipped tests that are left when this PR is merged and tag the appropriate people to get them un-skipped again ASAP.

Commit:
Fixing replace_panel_action.test.tsx

Commit:
Removing unnecessary comment

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13133/)
* Commit: ef3ede2a54e33026824720b3edd9f9c9091a73b1

watson commented on src/plugins/kibana_react/public/saved_objects/saved_object_finder.test.tsx:
```diff
@@ -629,7 +629,7 @@ describe('SavedObjectsFinder', () => {
 
   describe('loading state', () => {
-    it('should display a spinner during initial loading', () => {
+    it.skip('should display a spinner during initial loading', () => {
```
Here's the stack trace of the unhandled promise rejection:

```
TypeError: Cannot read property 'savedObjects' of undefined
        at SavedObjectFinder._lodash.default.debounce (./src/plugins/kibana_react/public/saved_objects/saved_object_finder.tsx:46:32)
        at process._tickCallback (internal/process/next_tick.js:68:7)
```

streamich commented:
@sqren I think it is pretty clear what this PR does.

We had a Zoom with @watson and agreed that we will not merge this with skipped tests, but @watson will create an issue and @elastic/kibana-app-arch team will fix the affected code or tests first.

//cc @stacey-gammon @ppisljar

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13146/)
* Commit: 0c786b3adcdbb483f3d06974052c7bb976f00886

watson commented on src/plugins/data/public/ui/query_string_input/query_bar_top_row.test.tsx:
```diff
@@ -117,7 +117,7 @@ describe('QueryBarTopRowTopRow', () => {
   });
 
-  it('Should render the given query', () => {
+  it.skip('Should render the given query', () => {
```
Here's the stack trace of the unhandled promise rejection:

```
TypeError: Cannot read property 'autocomplete' of undefined
    at QueryStringInputUI._defineProperty [as getSuggestions] (./src/plugins/data/public/ui/query_string_input/query_string_input.tsx:89:55)
    at QueryStringInputUI._defineProperty (./src/plugins/data/public/ui/query_string_input/query_string_input.tsx:141:39)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

watson replied:
It's the same error for the other skipped tests in this file as well

watson replied:
@lizozom I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson replied:
@lukeelmers @streamich I just saw that Liza is on vacation this week, would any of you be able to help me fix these tests in this PR?

We previously discussed fixing it in a separate PR, but I think it would speed things up if you could push a commit directly to this PR so we don't block it. Any help would be much appreciated 😃

watson commented on src/plugins/data/public/ui/query_string_input/query_string_input.test.tsx:
```diff
@@ -87,7 +87,7 @@ function wrapQueryStringInputInContext(testProps: any, storage?: any) {
 }
 
-describe('QueryStringInput', () => {
+describe.skip('QueryStringInput', () => {
```
All the tests in this `describe` block fail with an unhandled promise rejection. The stack traces are all variations of this one with different call-stacks leading up to it:

```
TypeError: Cannot read property 'autocomplete' of undefined
    at QueryStringInputUI._defineProperty [as getSuggestions] (./src/plugins/data/public/ui/query_string_input/query_string_input.tsx:89:55)
    at QueryStringInputUI._defineProperty (./src/plugins/data/public/ui/query_string_input/query_string_input.tsx:141:39)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

watson replied:
@lizozom I'm not sure if you're the right person to ping about this (blame `git blame` 😅), but I had to skip these tests as they were producing unhandled promise rejections due to an uncaught exception.

watson replied:
@lukeelmers @streamich I just saw that Liza is on vacation this week, would any of you be able to help me fix these tests in this PR?

We previously discussed fixing it in a separate PR, but I think it would speed things up if you could push a commit directly to this PR so we don't block it. Any help would be much appreciated 😃

jportner committed:
Update index_privileges

The prior commit got the test working but did not resolve the root
cause of the problem. Enzyme was throwing an exception because it
cannot handle wrapping a component that renders as an empty array
or null. Adding a Fragment wrapper around it fixed the issue.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13204/)
* Commit: c7dbf1b3ac75a6611cbdde40328a00fb72198089

watson committed:
un-skip fixed test

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13361/)
* Commit: 74ac5228138cd94a8bd4f99bfe9fcf64a4d1a2da

flash1293 committed:
re-enable saved object finder tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13383/)
* Commit: 8b98dcc46a5c60f238dee39acc621d625127e916

watson committed:
Remove use of deprecated API's not found by CI

streamich committed:
test: 💍 add missing data plugin mock in query_bar_top_row

streamich committed:
test: 💍 add missing mock in query_string_input tests

streamich commented on src/plugins/data/public/ui/query_string_input/__snapshots__/query_string_input.test.tsx.snap:
```diff
@@ -148,6 +148,74 @@ exports[`QueryStringInput Should disable autoFocus on EuiFieldText when disableA
               "setIsCollapsed": [MockFunction],
               "setIsVisible": [MockFunction],
             },
+            "data": Object {
```
This commit creates unnecessary snapshot of data service, but it is known issue and was already reported before: https://github.com/elastic/kibana/pull/51818#pullrequestreview-325066355

I'll create an issue to fix these mocks in a separate effort.

streamich replied:
https://github.com/elastic/kibana/issues/52181

streamich reviewed (approved):
App Architecture changes LGTM.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13414/)
* Commit: 63fdf69c30ccb50e001f2dafd4fd4c724281d9c8

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13434/)
* Commit: 5bb24810e918c27bc0a38b02444a580332a99b8e
